clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -0,0 +1,638 @@
+     * on the preference).
+     * @return Non-empty map of supported protocols and metadata
+     */
+    protected abstract LinkedHashMap<String, ByteBuffer> metadata();
+
+    /**
+     * Invoked when a group member has successfully joined a group.
+     * @param generation The generation that was joined
+     * @param memberId The identifier for the local member in the group
+     * @param protocol The protocol selected by the coordinator
+     * @param memberAssignment The assignment propagated from the group leader
+     */
+    protected abstract void onJoin(int generation,
+                                   String memberId,
+                                   String protocol,
+                                   ByteBuffer memberAssignment);
+
+    /**
+     * Perform synchronization for the group. This is used by the leader to push state to all the members
+     * of the group (e.g. to push partition assignments in the case of the new consumer)
+     * @param leaderId The id of the leader (which is this member)
+     * @param allMemberMetadata Metadata from all members of the group
+     * @return A map from each member to their state assignment
+     */
+    protected abstract Map<String, ByteBuffer> doSync(String leaderId,
+                                                      String protocol,
+                                                      Map<String, ByteBuffer> allMemberMetadata);
+
+    /**
+     * Invoked when the group is left (whether because of shutdown, metadata change, stale generation, etc.)
+     * @param generation The generation that was left
+     * @param memberId The identifier of the local member in the group
+     */
+    protected abstract void onLeave(int generation, String memberId);
+
+
+    /**
+     * Block until the coordinator for this group is known.
+     */
+    public void ensureCoordinatorKnown() {
+        while (coordinatorUnknown()) {
+            RequestFuture<Void> future = sendGroupMetadataRequest();
+            client.poll(future, requestTimeoutMs);
+
+            if (future.failed())
+                client.awaitMetadataUpdate();
+        }
+    }
+
+    /**
+     * Check whether the group should be rejoined (e.g. if metadata changes)
+     * @return true if it should, false otherwise
+     */
+    protected boolean needRejoin() {
+        return rejoinNeeded;
+    }
+
+    /**
+     * Reset the generation/memberId tracked by this member
+     */
+    public void resetGeneration() {
+        this.generation = OffsetCommitRequest.DEFAULT_GENERATION_ID;
+        this.memberId = JoinGroupRequest.UNKNOWN_MEMBER_ID;
+        rejoinNeeded = true;
+    }
+
+    /**
+     * Ensure that the group is active (i.e. joined and synced)
+     */
+    public void ensureActiveGroup() {
+        if (!needRejoin())
+            return;
+
+        // onLeave only invoked if we have a valid current generation
+        onLeave(generation, memberId);
```

Comment:
@hachikuji About (3) and (4), I think the confusion comes from we are reusing the term **Join** here while we already have a JoinGroupRequest. But I am also a little concerned about exposing the concept of **generation**. **Generation** is a purely implementation detail so it would be nice not to expose that to user.

My original thinking was `onRebalanceBegin()` and `onRebalanceEnd()`. It is a precise description and avoids reusing JoinGroup, but it exposes the concept of **rebalance** to user. That was why I switched to `onGroupJoinBegin` and `onGroupJoinEnd` - because it is very easy for user to understand. People who overrides those two methods don't need to have ideas about internal details such as JoinGroupRequest, Generation, Rebalance, etc. They only need to know:

1. **onGropuJoinStart** - Do something before taking action to join/rejoin the group.
2. **generateAssignments** - Run the customized resource assigning algorithm.
3. **onGroupJoinFinish** - Do something after you joined the group

From user's point of view they are only doing a Group Join (it has nothing to do with JoinGroupReqeust). Group Join looks more intuitive than names related to rebalance or generation.

I just want to reason a little bit on the names. Because those method names are essentially public API that user will override, I feel making them easy to understand is important.